File: worker/src/jobs/compliance/preparation.ts
// ComplianceGuard — input preparation. Faithful TS port of app/preparation.py.
import {
  type Cell, type Row,
  s, num0, round2, isValidGst, invoicePlace, groupBy, sumCol, findCol,
} from "./helpers";
import type { SheetData } from "./excel";
import type { Frame } from "./types";

/** Resolve the first candidate column that exists in the sheet, else null. */
function pick(columns: string[], candidates: string[]): string | null {
  for (const c of candidates) if (columns.includes(c)) return c;
  return null;
}

/** sales_df.get(col, default) over a row, given the resolved column (or null). */
function getv(row: Row, col: string | null, dflt: Cell = ""): Cell {
  return col ? row[col] ?? dflt : dflt;
}

// ---------------------------------------------------------------------------
// prepare_raw_sales_sheet — filters to valid-GST (B2B) rows only.
// ---------------------------------------------------------------------------
export function prepareRawSales(sales: SheetData): Frame {
  const cols = sales.columns;
  const cInvNo = pick(cols, ["InvoiceNo"]);
  const cQty = pick(cols, ["QTY"]);
  const cUnit = pick(cols, ["Unit Price"]);
  const cTaxPct = pick(cols, ["TaxPercent"]);
  const cTaxable = pick(cols, ["Taxable Value", "TaxableAmount"]);
  const cCgst = pick(cols, ["CGSTAmount"]);
  const cSgst = pick(cols, ["SGSTAmount"]);
  const cIgst = pick(cols, ["IGSTAmount"]);
  const cTax = pick(cols, ["TaxAmount"]);
  const cSelling = pick(cols, ["Selling Price"]);
  const cDisc = pick(cols, ["Discount on item"]);
  const cDate = pick(cols, ["DateofInvoice", "Date of Invoice"]);
  const cOrderNo = pick(cols, ["ExternalOrderNo", "External Order No"]);
  const cChannel = pick(cols, ["Order Channel"]);
  const cCustomer = pick(cols, ["Billing Name of Customer", "Name of Customer"]);
  const cState = pick(cols, ["State"]);
  const cSku = pick(cols, ["ClientSKU", "ClientSku"]);
  const cDesc = pick(cols, ["DescriptionofGoods", "DescriptionOfGoods"]);
  const cHsn = pick(cols, ["HSN Code"]);
  const cGst = pick(cols, ["GST No"]);
  const cIrn = pick(cols, ["E-invoice No.(IRN)"]);
  const cEway = pick(cols, ["E-way bill number"]);

  const out: Row[] = [];
  for (const r of sales.rows) {
    if (!isValidGst(r[cGst ?? "GST No"])) continue;
    const taxable = num0(getv(r, cTaxable, 0));
    const totalTax = num0(getv(r, cTax, 0));
    out.push({
      "Invoice No": s(getv(r, cInvNo, "")),
      "Invoice Place": invoicePlace(getv(r, cInvNo, "")),
      "Invoice Date": getv(r, cDate, ""),
      "Order No": getv(r, cOrderNo, ""),
      Channel: getv(r, cChannel, ""),
      "Billing Name of Customer": getv(r, cCustomer, ""),
      State: getv(r, cState, ""),
      "Client SKU": getv(r, cSku, ""),
      "Description of Goods": getv(r, cDesc, ""),
      "HSN Code": getv(r, cHsn, ""),
      "GST No": getv(r, cGst, ""),
      "E-invoice No.(IRN)": getv(r, cIrn, ""),
      "E-way bill number": getv(r, cEway, ""),
      QTY: num0(getv(r, cQty, 0)),
      "Unit Price": num0(getv(r, cUnit, 0)),
      TaxPercent: num0(getv(r, cTaxPct, 0)),
      "Taxable Value": taxable,
      CGST: num0(getv(r, cCgst, 0)),
      SGST: num0(getv(r, cSgst, 0)),
      IGST: num0(getv(r, cIgst, 0)),
      "Total Tax": totalTax,
      "Selling Price": num0(getv(r, cSelling, 0)),
      Discount: num0(getv(r, cDisc, 0)),
      Total: round2(taxable + totalTax),
    });
  }

  // Invoice Total = groupby('Invoice No')['Total'].transform('sum')
  const groups = groupBy(out, (r) => s(r["Invoice No"]));
  const invTotal = new Map<string, number>();
  for (const [k, rows] of groups) invTotal.set(k, sumCol(rows, "Total"));
  for (const r of out) r["Invoice Total"] = invTotal.get(s(r["Invoice No"])) ?? 0;

  const columns = [
    "Invoice No", "Invoice Place", "Invoice Date", "Order No", "Channel",
    "Billing Name of Customer", "State", "Client SKU", "Description of Goods",
    "HSN Code", "GST No", "E-invoice No.(IRN)", "E-way bill number", "QTY",
    "Unit Price", "TaxPercent", "Taxable Value", "CGST", "SGST", "IGST",
    "Total Tax", "Selling Price", "Discount", "Total", "Invoice Total",
  ];
  return { columns, rows: out };
}

// ---------------------------------------------------------------------------
// prepare_einvoice_data — valid-GST filter + dedup by invoice number.
// ---------------------------------------------------------------------------
export function prepareEinvoice(einv: SheetData): Frame {
  let rows = einv.rows;
  const cols = einv.columns;

  const gstinCol = findCol(cols, ["gstin/uin", "gstin of recipient", "recipient gstin"]);
  if (gstinCol) rows = rows.filter((r) => isValidGst(r[gstinCol]));

  const invNoCol = findCol(cols, ["invoice number", "invoice no", "doc no"]);
  const invValCol = findCol(cols, ["invoice value", "inv value", "total value"]);
  const irnCol = findCol(cols, ["irn"]);
  const dateCol = findCol(cols, ["invoice date", "inv date", "doc date"]);
  const gstin2 = findCol(cols, ["gstin/uin", "gstin of recipient", "recipient gstin"]);
  const nameCol = findCol(cols, ["receiver name", "trade name", "legal name", "party name"]);

  if (!invNoCol) throw new Error(`Cannot find Invoice Number column. Available: ${cols.slice(0, 15).join(", ")}`);
  if (!invValCol) throw new Error(`Cannot find Invoice Value column. Available: ${cols.slice(0, 15).join(", ")}`);

  const seen = new Set<string>();
  const out: Row[] = [];
  for (const r of rows) {
    const key = s(r[invNoCol]);                 // drop_duplicates(keep='first') is on raw value
    const rawKey = String(r[invNoCol] ?? "");
    if (seen.has(rawKey)) continue;
    seen.add(rawKey);
    out.push({
      "E-Invoice No": key,
      "E-Invoice Amount": num0(r[invValCol]),
      IRN: irnCol ? s(r[irnCol]) : "",
      "Invoice date": dateCol ? r[dateCol] ?? "" : "",
      "GSTIN/UIN of Recipient": gstin2 ? s(r[gstin2]) : "",
      "Receiver Name": nameCol ? s(r[nameCol]) : "",
    });
  }
  return {
    columns: ["E-Invoice No", "E-Invoice Amount", "IRN", "Invoice date", "GSTIN/UIN of Recipient", "Receiver Name"],
    rows: out,
  };
}

// ---------------------------------------------------------------------------
// prepare_eway_data — normalise Doc.No / Other Party GSTIN, keep all columns.
// ---------------------------------------------------------------------------
export function prepareEway(eway: SheetData): Frame {
  const cols = [...eway.columns];
  const docCol = findCol(cols, ["doc.no", "doc no", "invoice no", "inv no"]);
  const gstinCol = findCol(cols, ["other party gstin", "to gstn", "from gstn", "party gstin"]);

  const rename = new Map<string, string>();
  if (docCol && docCol !== "Doc.No") rename.set(docCol, "Doc.No");
  if (gstinCol && gstinCol !== "Other Party GSTIN") rename.set(gstinCol, "Other Party GSTIN");

  const newCols = cols.map((c) => rename.get(c) ?? c);
  if (!newCols.includes("Doc.No")) {
    throw new Error(`Cannot find Doc/Invoice No column in E-way Bill. Available: ${cols.slice(0, 15).join(", ")}`);
  }

  const rows: Row[] = eway.rows.map((r) => {
    const o: Row = {};
    for (const c of cols) o[rename.get(c) ?? c] = r[c] ?? null;
    o["Doc.No"] = s(o["Doc.No"]);
    return o;
  });
  return { columns: newCols, rows };
}

File: dentalkart-gst-suite-master-UpdatedCodeNiranjanRaised/dentalkart-gst-suite-master/app/preparation.py
import numpy as np
import pandas as pd
from app.helpers import is_valid_gst, is_valid_gst_vectorized, get_invoice_place
from app.config import INVOICE_PLACE_MAPPING


def prepare_raw_sales_sheet(sales_df):
    """Prepare raw sales sheet - FILTERS FOR VALID GST ONLY (B2B)"""
    # Vectorized GST filter (was: .apply(is_valid_gst) - row by row)
    sales_df = sales_df[is_valid_gst_vectorized(sales_df['GST No'])]

    inv_no_str = sales_df['InvoiceNo'].astype(str).str.strip()
    inv_prefixes = inv_no_str.str[:3].str.upper()

    # Batch all numeric conversions at once using a dict
    num_cols = {
        'QTY': sales_df.get('QTY', 0),
        'Unit Price': sales_df.get('Unit Price', 0),
        'TaxPercent': sales_df.get('TaxPercent', 0),
        'Taxable Value': sales_df.get('Taxable Value', sales_df.get('TaxableAmount', 0)),
        'CGST': sales_df.get('CGSTAmount', 0),
        'SGST': sales_df.get('SGSTAmount', 0),
        'IGST': sales_df.get('IGSTAmount', 0),
        'Total Tax': sales_df.get('TaxAmount', 0),
        'Selling Price': sales_df.get('Selling Price', 0),
        'Discount': sales_df.get('Discount on item', 0),
    }
    numeric_data = {}
    for col_name, col_data in num_cols.items():
        numeric_data[col_name] = pd.to_numeric(col_data, errors='coerce').fillna(0)

    raw = pd.DataFrame({
        'Invoice No': inv_no_str,
        'Invoice Place': inv_prefixes.map(INVOICE_PLACE_MAPPING).fillna('Unknown'),
        'Invoice Date': sales_df.get('DateofInvoice', sales_df.get('Date of Invoice', '')),
        'Order No': sales_df.get('ExternalOrderNo', sales_df.get('External Order No', '')),
        'Channel': sales_df.get('Order Channel', ''),
        'Billing Name of Customer': sales_df.get('Billing Name of Customer', sales_df.get('Name of Customer', '')),
        'State': sales_df.get('State', ''),
        'Client SKU': sales_df.get('ClientSKU', sales_df.get('ClientSku', '')),
        'Description of Goods': sales_df.get('DescriptionofGoods', sales_df.get('DescriptionOfGoods', '')),
        'HSN Code': sales_df.get('HSN Code', ''),
        'GST No': sales_df.get('GST No', ''),
        'E-invoice No.(IRN)': sales_df.get('E-invoice No.(IRN)', ''),
        'E-way bill number': sales_df.get('E-way bill number', ''),
        **numeric_data,
    })

    raw['Total'] = (raw['Taxable Value'] + raw['Total Tax']).round(2)
    raw['Invoice Total'] = raw.groupby('Invoice No')['Total'].transform('sum')
    return raw


def prepare_einvoice_data(einv_df):
    """Prepare E-Invoice data - filters for valid GST, deduplicates by Invoice Number"""
    gstin_col = None
    for c in einv_df.columns:
        if any(kw in str(c).lower() for kw in ['gstin/uin', 'gstin of recipient', 'recipient gstin']):
            gstin_col = c
            break
    if gstin_col:
        # Vectorized GST filter
        einv_df = einv_df[is_valid_gst_vectorized(einv_df[gstin_col])]

    col_lower = {c: str(c).lower().strip() for c in einv_df.columns}
    def find_col(keywords):
        for col, lc in col_lower.items():
            if any(kw in lc for kw in keywords):
                return col
        return None

    inv_no_col  = find_col(['invoice number', 'invoice no', 'doc no'])
    inv_val_col = find_col(['invoice value', 'inv value', 'total value'])
    irn_col     = find_col(['irn'])
    date_col    = find_col(['invoice date', 'inv date', 'doc date'])
    gstin_col2  = find_col(['gstin/uin', 'gstin of recipient', 'recipient gstin'])
    name_col    = find_col(['receiver name', 'trade name', 'legal name', 'party name'])

    if not inv_no_col:
        raise ValueError(f"Cannot find Invoice Number column. Available: {list(einv_df.columns)[:15]}")
    if not inv_val_col:
        raise ValueError(f"Cannot find Invoice Value column. Available: {list(einv_df.columns)[:15]}")

    einv_u = einv_df.drop_duplicates(subset=[inv_no_col], keep='first')
    out = pd.DataFrame()
    out['E-Invoice No']     = einv_u[inv_no_col].astype(str).str.strip()
    out['E-Invoice Amount'] = pd.to_numeric(einv_u[inv_val_col], errors='coerce').fillna(0)
    out['IRN']              = einv_u[irn_col].astype(str).str.strip() if irn_col else ''
    out['Invoice date']     = einv_u[date_col] if date_col else ''
    out['GSTIN/UIN of Recipient'] = einv_u[gstin_col2].astype(str).str.strip() if gstin_col2 else ''
    out['Receiver Name']    = einv_u[name_col].astype(str).str.strip() if name_col else ''
    return out


def prepare_eway_data(eway_df):
    """Normalize E-way Bill columns."""
    col_lower = {str(c).lower().strip(): c for c in eway_df.columns}
    def find_col(keywords):
        for lc, orig in col_lower.items():
            if any(kw in lc for kw in keywords):
                return orig
        return None
    doc_col = find_col(['doc.no', 'doc no', 'invoice no', 'inv no'])
    gstin_col = find_col(['other party gstin', 'to gstn', 'from gstn', 'party gstin'])
    if doc_col and doc_col != 'Doc.No':
        eway_df = eway_df.rename(columns={doc_col: 'Doc.No'})
    if gstin_col and gstin_col != 'Other Party GSTIN':
        eway_df = eway_df.rename(columns={gstin_col: 'Other Party GSTIN'})
    if 'Doc.No' not in eway_df.columns:
        raise ValueError(f"Cannot find Doc/Invoice No column in E-way Bill. Available: {list(eway_df.columns)[:15]}")
    eway_df['Doc.No'] = eway_df['Doc.No'].astype(str).str.strip()
    return eway_df
